### ISSUE TITLE
'make oct' overrides LFLAGS, cannot find -loctinterp

### ISSUE THREAD

**Issue by mtmiller** [fangq/mmc#18]
The `make oct` target sets the environment variable `LFLAGS`. If this variable is set by the user, it overrides the built in link options that are necessary for finding Octave's libraries.

Commands called by mkoctfile without any variables set:

```
$ mkoctfile -v foo.cc
g++ -c  -fPIC -I/opt/gnu/octave/include/octave-4.4.1/octave/.. -I/opt/gnu/octave/include/octave-4.4.1/octave -I/opt/gnu/octave/include  -pthread -fopenmp -g -O2    foo.cc -o foo.o
g++ -I/opt/gnu/octave/include/octave-4.4.1/octave/.. -I/opt/gnu/octave/include/octave-4.4.1/octave -I/opt/gnu/octave/include  -pthread -fopenmp -g -O2 -shared -Wl,-Bsymbolic  -o foo.oct  foo.o   -L/opt/gnu/octave/lib/octave/4.4.1 -L/opt/gnu/octave/lib -loctinterp -loctave  
```

Commands called by mkoctfile with LFLAGS set:

```
$ LFLAGS=-g mkoctfile -v foo.cc
g++ -c  -fPIC -I/opt/gnu/octave/include/octave-4.4.1/octave/.. -I/opt/gnu/octave/include/octave-4.4.1/octave -I/opt/gnu/octave/include  -pthread -fopenmp -g -O2    foo.cc -o foo.o
g++ -I/opt/gnu/octave/include/octave-4.4.1/octave/.. -I/opt/gnu/octave/include/octave-4.4.1/octave -I/opt/gnu/octave/include  -pthread -fopenmp -g -O2 -shared -Wl,-Bsymbolic  -o foo.oct  foo.o   -g -loctinterp -loctave  
```

The first command passes `-L/opt/gnu/octave/lib/octave/4.4.1 -L/opt/gnu/octave/lib` to the linker, these options are necessary for finding the right copy of the Octave libraries installed on the system.

Originally reported at https://github.com/flathub/org.octave.Octave/issues/33.

**Comment by koenigjaeger** [comment]
So the link options have to be set in the make file (under mmc-master/src)? Or is there a way to do this temporary through octave? i.e. like system("make oct")

**Comment by mtmiller** [comment]
@koenigjaeger this can be fixed in the makefile. But in the meantime, you can try building with

    system ("make oct OPENMP='-L/app/lib/octave/4.4.1'");

since the makefile passes whatever the contents of the `OPENMP` variable are to the linker.

**Comment by koenigjaeger** [comment]
Here is the following output...

```
>> system ("make oct OPENMP='-L/app/lib/octave/4.4.1'");
Building bin/mmc
CC=cc CXX=g++ LFLAGS='-L/app/lib/octave/4.4.1' LDFLAGS='' CPPFLAGS='-c -Wall -g -fno-strict-aliasing -m64 -DMMC_USE_SSE -DHAVE_SSE2 -msse -msse2 -msse3 -mssse3 -msse4.1 -O3 -L/app/lib/octave/4.4.1  -fPIC -DMCX_CONTAINER -DUSE_OS_TIMER -std=c++11'  mkoctfile --mex mmclab.cpp -I../src -o bin/mmc built/posix_randr.o built/simpmesh.o built/tettracing.o built/mcx_utils.o built/tictoc.o built/mmc.o built/mmc_host.o built/cjson/cJSON.o -o ../mmclab/mmc
mmclab.cpp: In function ‘void mexFunction(int, mxArray**, int, const mxArray**)’:
mmclab.cpp:93:14: warning: variable ‘usewaitbar’ set but not used [-Wunused-but-set-variable]
   int        usewaitbar=1;
              ^~~~~~~~~~
mmclab.cpp: In function ‘void mmc_set_field(const mxArray*, const mxArray*, int, mcconfig*, tetmesh*)’:
mmclab.cpp:567:67: warning: format ‘%d’ expects argument of type ‘int’, but argument 2 has type ‘dimtype {aka long int}’ [-Wformat=]
         printf("mmc.srcpattern=[%d %d];\n",arraydim[0],arraydim[1]);
                                                                   ^
mmclab.cpp:567:67: warning: format ‘%d’ expects argument of type ‘int’, but argument 3 has type ‘dimtype {aka long int}’ [-Wformat=]
>>
```

**Comment by koenigjaeger** [comment]
I see that its just a warning so it shouldn't be an issue.

**Comment by koenigjaeger** [comment]
Although it seems like it builds to bin/mmc and it doesn't look like any was created.

**Comment by mtmiller** [comment]
@koenigjaeger the result is the file mmclab/mmc.mex.

**Comment by koenigjaeger** [comment]
@mtmiller Sorry I though mkoct outputted with the file type of *.oct. Thank you for your help! I'll test the toolbox and report anymore issues.

**Comment by mtmiller** [comment]
@koenigjaeger By default, yes mkoctfile compiles to a .oct file. But this project uses the mex API, and `mkoctfile --mex` is used to link against Octave's implementation of the mex API.

**Comment by koenigjaeger** [comment]
Wouldn't it be better to use *.oct? See: https://octave.org/doc/interpreter/Mex_002dFiles.html; where is mentions "...a mex-file can never have the same performance in Octave as the equivalent oct-file."

**Comment by mtmiller** [comment]
If you want to encourage the authors here to rewrite their code to use the Octave API, or attempt to do that work for them, that might be something to look at. Octave offers an implementation of the mex API for existing projects that don't want to have to do that porting effort and maintain two sets of interfaces for Matlab and Octave.

**Comment by fangq** [comment]
@koenigjaeger and @mtmiller, there is no speed impact to convert mmclab/mmclab using oct APIs. The actual computing part is compiled separately by gcc (mmclab) or nvcc (mcxlab) and is only linked to the generated mex file in the end, therefore, the computing speed in either mmclab or mcxlab should be nearly identical in matlab/octave, which is also identical to their corresponding native binaries (mmc.exe, mcx.exe). 

The API only helps passing in/out variables from octave/matlab, which has a very small overhead in both simulations. Using mex API for both matlab/octave simplifies code maintenance.

@koenigjaeger, add path to mmclab/ and try one of the demo scripts in mmclab/example (you need to install iso2mesh to run most examples).

I will also read about flatpak. wanted to create standard packages for mcx/mmc/mcxlab/mmclab on Ubuntu/Fedora, but only managed to [create deb/rpm packages](https://github.com/fangq/mcx/blob/master/Makefile#L23-L31).

**Comment by koenigjaeger** [comment]
@fangq It would be great if there was a flatpak. It could be especially useful for MCXstudio with most flatpaks distributed on https://flathub.org/home. I've always meant to learn how to build flatpaks myself, so once I get some time I'll check out building MCXstudio for flatpak and let you know if I make any progress.

**Comment by fangq** [comment]
@koenigjaeger, I am still trying to figure out some dependency issues for mcxstudio if compiled on a packaging server. it needs lazarus (lazbuild) and a 3rd party component (glscene), and the latter is not generally available on Linux distributions. but I think I am getting close, here is the thread of discussions

http://forum.lazarus.freepascal.org/index.php/topic,42438.0.html

the mmc/mmclab/mcx/mcxlab/mcxcl/mcxlabcl packages should be more straightforward (and has mostly been done for deb). Just need some review and sponsorship from debian package maintainers. I saw @mtmiller is a DD, wondering if you mind me approaching you to learn how to bring these packages to debian? 

mcx needs CUDA, don't know how these proprietary packages are handled in debian these days.

**Comment by mtmiller** [comment]
Yeah feel free to email me about Debian packaging, can't promise anything but I will help with what I can.

Debian non-free does have CUDA packages available, any package that requires CUDA would also go into the non-free repository, unless it can do some kind of runtime detection and fully function without it.

**Comment by fangq** [comment]
hi @mtmiller, I recently [resumed my package maintainer activities](https://fedoraproject.org/wiki/User:Fangq) for Fedora, and started making progress on packaging a list of my research software (most of them are matlab/octave toolboxes, like iso2mesh, zmat, jsonlab, mcxlabcl, jnifti, ...), some of the packages that I am working on are listed here

https://github.com/fangq/fedorapkg/branches

next, I would like to learn how to contribute these packages to Debian. I am happy either assisting someone to create these packages, or applying for DM and create/maintain these packages myself. 

from reading this page 

https://www.debian.org/doc/manuals/developers-reference/ch02.html

to become a new DM, I need a sponsor, I am wondering if you mind being my sponsor?

**Comment by mtmiller** [comment]
> I would like to learn how to contribute these packages to Debian. I am happy either assisting someone to create these packages, or applying for DM and create/maintain these packages myself.
> 
> from reading this page
> 
> https://www.debian.org/doc/manuals/developers-reference/ch02.html
> 
> to become a new DM, I need a sponsor, I am wondering if you mind being my sponsor?

Maintaining packages within teams in Debian is strongly recommended. I would recommend that you consider maintaining these packages within the [Debian Science](https://wiki.debian.org/DebianScience) team. I follow this team's activity, and it appears to me that there are several active sponsors and a lot of accumulated wisdom about maintaining scientific packages. So I would recommend starting packaging repositories for those packages you want to work on in Debian, file ITP bugs, and then contact the Science team for help and sponsorship.

- https://wiki.debian.org/DebianScience
- https://wiki.debian.org/DebianScience/Sponsoring
- https://wiki.debian.org/ITP
- https://www.debian.org/doc/manuals/developers-reference/ch05.en.html#new-packages

**Comment by fangq** [comment]
great! thank you for the pointers, DebianScience does look like the right place for these packages. I will read the documentation and start working on the initial packaging.
